services/sts/src/main/java/software/amazon/awssdk/services/sts/internal/StsWebIdentityCredentialsProviderFactory.java

```diff
@@ -0,0 +1,135 @@
+
+            this.stsClient = StsClient.builder()
+                                      .credentialsProvider(AnonymousCredentialsProvider.create())
+                                      .overrideConfiguration(o -> o.retryPolicy(r -> r.retryCondition(retryCondition)))
+                                      .build();
+
+            AssumeRoleWithWebIdentityRequest request = AssumeRoleWithWebIdentityRequest.builder()
+                                                                                       .roleArn(credentialProperties.roleArn())
+                                                                                       .roleSessionName(sessionName)
+                                                                                       .build();
+
+            AssumeRoleWithWebIdentityRequestSupplier supplier =
+                new AssumeRoleWithWebIdentityRequestSupplier(request,
+                                                             credentialProperties.webIdentityTokenFilePath());
+
+            this.credentialsProvider =
+                StsAssumeRoleWithWebIdentityCredentialsProvider.builder()
+                                                               .stsClient(stsClient)
+                                                               .refreshRequest(supplier)
+                                                               .build();
+        }
+
+        @Override
+        public AwsCredentials resolveCredentials() {
+            return this.credentialsProvider.resolveCredentials();
+        }
+
+        @Override
+        public void close() {
+            IoUtils.closeQuietly(credentialsProvider, null);
+            IoUtils.closeQuietly(stsClient, null);
+        }
+    }
+
+    private static final class AssumeRoleWithWebIdentityRequestSupplier implements Supplier {
+
+        private final AssumeRoleWithWebIdentityRequest request;
+        private final String webIdentityTokenFilePath;
+
+        AssumeRoleWithWebIdentityRequestSupplier(AssumeRoleWithWebIdentityRequest request,
+                                                 String webIdentityTokenFilePath) {
+            this.request = request;
+            this.webIdentityTokenFilePath = webIdentityTokenFilePath;
+        }
+
+        @Override
+        public Object get() {
+            return request.toBuilder().webIdentityToken(getToken(webIdentityTokenFilePath)).build();
+        }
+
+        private String getToken(String filePath) {
+            try (InputStream webIdentityTokenStream =
+                     FunctionalUtils.invokeSafely(() -> Files.newInputStream(Paths.get(filePath)))) {
+                return FunctionalUtils.invokeSafely(() -> IoUtils.toUtf8String(webIdentityTokenStream));
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+
+    private static final class StsRetryCondition implements RetryCondition {
```

Comment:
Should this be included in the other STS credential providers? Maybe the entire STS retry condition?

Reply:
I can rename but this is specific to this provider.

Reply:
I more meant, should we do this for all STS usages - retry on this exception.

Reply:
AFAIK, this is specific to this use case (was called out in SEP). Can see what other SDKs are doing here though.